File: 48-rotate-image/48-rotate-image.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        for k in range(n//2):
            t1 = tuple(list(matrix[k][i+k] for i in range(n-2*k)))
            t2 = tuple(list(matrix[i+k][-1-k] for i in range(n-2*k)))
            t3 = tuple(list(matrix[-1-k][-(i+1+k)] for i in range(n-2*k)))
            t4 = tuple(list(matrix[-(i+1+k)][k] for i in range(n-2*k)))
            print(t1)
            print(t2)
            print(t3)
            print(t4)
            for i in range(n-2*k):
                matrix[k][i+k] = t4[i]
                matrix[i+k][-1-k] = t1[i]
                matrix[-1-k][-(i+1+k)] = t2[i]
                matrix[-(i+1+k)][k] = t3[i]

File: 48-rotate-image/48-rotate-image.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int n = matrix.size();
        vector<int> t1;
        vector<int> t2;
        vector<int> t3;
        vector<int> t4;
        for (int k=0; k<(n/2); k++) {
            for (int i=0; i<(n-2*k); i++) {
                t1.push_back(matrix[k][i+k]);
            }
            for (int i=0; i<(n-2*k); i++) {
                t2.push_back(matrix[i+k][n-1-k]);
            }
            for (int i=0; i<(n-2*k); i++) {
                t3.push_back(matrix[n-1-k][n-(i+1+k)]);
            }
            for (int i=0; i<(n-2*k); i++) {
                t4.push_back(matrix[n-(i+1+k)][k]);
            }
            for (int i=0; i<(n-2*k); i++) {
                matrix[k][i+k] = t4[i];
                matrix[i+k][n-1-k] = t1[i];
                matrix[n-1-k][n-(i+1+k)] = t2[i];
                matrix[n-(i+1+k)][k] = t3[i];
            }
            t1.erase(t1.begin(), t1.end());
            t2.erase(t2.begin(), t2.end());
            t3.erase(t3.begin(), t3.end());
            t4.erase(t4.begin(), t4.end());
        }
    }
};